File: training-data/ocaml/longest_palindromic_substring.ml
let expand_from_center s l r =
  let l = ref l and r = ref r in
  let n = String.length s in
  while !l >= 0 && !r < n && s.[!l] = s.[!r] do
    decr l;
    incr r
  done;
  !r - !l - 1

let longest_palindrome s =
  let n = String.length s in
  if n = 0 then ""
  else begin
    let start = ref 0 and max_len = ref 1 in
    for i = 0 to n - 1 do
      let len1 = expand_from_center s i i in
      let len2 = expand_from_center s i (i + 1) in
      let len = max len1 len2 in
      if len > !max_len then begin
        max_len := len;
        start := i - ((len - 1) / 2)
      end
    done;
    String.sub s !start !max_len
  end

let () =
  print_endline (longest_palindrome "babad");
  print_endline (longest_palindrome "cbbd")

File: training-data/nim/longest_palindromic_substring.nim
proc longestPalindrome(s: string): string =
  if s.len == 0:
    return ""
  var start = 0
  var maxLen = 1
  proc expand(l, r: int): int =
    var l = l
    var r = r
    while l >= 0 and r < s.len and s[l] == s[r]:
      l -= 1
      r += 1
    r - l - 1
  for i in 0 ..< s.len:
    let len1 = expand(i, i)
    let len2 = expand(i, i + 1)
    let length = max(len1, len2)
    if length > maxLen:
      maxLen = length
      start = i - (length - 1) div 2
  s[start ..< start + maxLen]

echo longestPalindrome("babad")
echo longestPalindrome("cbbd")
